Sync/Synchronizers/IndependentRecordSynchronizer.swift

```diff
@@ -120,4 +130,36 @@ extension TimestampedSingleCollectionSynchronizer {
                 return deferMaybe(timestamp)
             }
     }
+
+    func uploadRecordsSingleBatch<T>(_ records: [Record<T>], lastTimestamp: Timestamp, storageClient: Sync15CollectionClient<T>) -> DeferredTimestamp {
+        if records.isEmpty {
+            log.debug("No modified records to upload.")
+            return deferMaybe(lastTimestamp)
+        }
+
+        if records.count > storageClient.maxBatchPostRecords {
+            return deferMaybe(TooManyRecordsError())
+        }
+
+        var additionalPosts = 0
+        func onCollectionUploaded(result: POSTResult, timestamp: Timestamp?) -> DeferredTimestamp {
+            additionalPosts += 1
+            return deferMaybe(timestamp ?? 0)
+        }
+
+        let batch = storageClient.newBatch(ifUnmodifiedSince: (lastTimestamp == 0) ? nil : lastTimestamp, onCollectionUploaded: onCollectionUploaded)
+        return batch.addRecords(records)
+            >>== { _ in
+                return batch.flush()
```

Comment:
If the batch API is disabled, this `flush` results in the records being uploaded to the server. It's not helpful for us then to throw `TooManyRecordsError`, right? The cat is out of the bag.

What we want is:

- To increase our chances of success by limiting ourselves to the size of a single POST.
- To upload in no more than a single commit, or no more than a single request, depending on how optimistic we are. The uploader should help us with that.
- If we supply too many records to the uploader, fail _then_, not after some have been uploaded! (This is why it's nice to push the count down into the uploader.)
- Regardless of whether batching was enabled, use the list of successfully uploaded records as our source of truth moving on — drop them from `local` and add them as children of `mobile`. If there was some problem with some of the records, _and_ the batching API was disabled, this will leave us slightly out of sync with the server, but in a more recoverable and incremental state than simply failing.

Thoughts?